src/vset.c

```diff
@@ -0,0 +1,1249 @@
+void *pvGet(pVector *vec, uint32_t idx) {
+    if (!vec || idx >= vec->len) return NULL;
+    return vec->data[idx];
+}
+
+/* Frees the memory used by the pVector.
+ *
+ * Arguments:
+ *   sv - The pVector to free.
+ *
+ * Return:
+ *   None. */
+void pvFree(pVector *sv) {
+    if (sv) zfree(sv);
+}
+
+uint32_t pvFind(pVector *sv, void *elem) {
+    if (!sv || sv->len == 0) return 0;
+
+    for (uint32_t i = 0; i < sv->len; i++) {
+        if (sv->data[i] == elem) {
+            return i;
+        }
+    }
+    return sv->len;
+}
+/*************************************************************************************************************
+ *                                pVector End
+ *************************************************************************************************************/
+#define VSET_BUCKET_NONE -1      // matching the NULL case
+#define VSET_BUCKET_SINGLE 0x1UL // xx1 (assuming sds)
+#define VSET_BUCKET_VECTOR 0x2UL // 010
+#define VSET_BUCKET_HT 0x4UL     // 100
+#define VSET_BUCKET_RAX 0x6UL    // 110
+
+#define VSET_TAG_MASK 0x7UL
+#define VSET_PTR_MASK (~VSET_TAG_MASK)
+
+// Determine bucket type
+static inline int vsetBucketType(vsetBucket *b) {
+    if (b == NULL) return VSET_BUCKET_NONE;
+
+    uintptr_t bits = (uintptr_t)b;
+    if (bits & 0x1)
+        return VSET_BUCKET_SINGLE;
+    return bits & VSET_TAG_MASK;
+}
+
+// Access raw pointer
+static inline void *vsetBucketRawPtr(vsetBucket *b) {
+    return (void *)((uintptr_t)b & VSET_PTR_MASK);
+}
+
+// Accessors with type assertions
+static inline pVector *vsetBucketVector(vsetBucket *b) {
+    assert(vsetBucketType(b) == VSET_BUCKET_VECTOR);
+    return (pVector *)vsetBucketRawPtr(b);
+}
+
+static inline hashtable *vsetBucketHashtable(vsetBucket *b) {
+    assert(vsetBucketType(b) == VSET_BUCKET_HT);
+    return (hashtable *)vsetBucketRawPtr(b);
+}
+
+static inline rax *vsetBucketRax(vsetBucket *b) {
+    assert(vsetBucketType(b) == VSET_BUCKET_RAX);
+    return (rax *)vsetBucketRawPtr(b);
+}
+
+static inline void *vsetBucketSingle(vsetBucket *b) {
+    return b;
+}
+
+// Setters
+static inline vsetBucket *vsetBucketSetType(vsetBucket *b, int type) {
+    uintptr_t p = (uintptr_t)b;
+    return (vsetBucket *)(p | (type & VSET_TAG_MASK));
+}
+
+static inline vsetBucket *vsetBucketSetVector(vsetBucket *b, pVector *vec) {
+    UNUSED(b);
+    return vsetBucketSetType(vec, VSET_BUCKET_VECTOR);
+}
+
+static inline vsetBucket *vsetBucketSetHashtable(vsetBucket *b, hashtable *ht) {
+    UNUSED(b);
+    return vsetBucketSetType(ht, VSET_BUCKET_HT);
+}
+
+static inline vsetBucket *vsetBucketSetSingle(vsetBucket *b, void *ptr) {
+    UNUSED(b);
+    return ptr;
+}
+
+static inline vsetBucket *vsetBucketSetNone(vsetBucket *b) {
+    UNUSED(b);
+    return NULL;
+}
+
+static inline vsetBucket *vsetBucketSetRax(vsetBucket *b, rax *r) {
+    UNUSED(b);
+    return vsetBucketSetType(r, VSET_BUCKET_RAX);
+}
```

Comment:
I wouldn't call these setters. They are constructors, I guess. How about From? (as in ValkeyModule_CreateStringFromLongLong, etc.)

The ignored `b` is just confusing. Let's skip it. (If I pass in a bucket and get one back, I would expect the passed in `b` to be freed or converted, not just ignored.)

```suggestion
static inline vsetBucket *vsetBucketFromVector(pVector *vec) {
    return vsetBucketFromRawPtr(vec, VSET_BUCKET_VECTOR);
}

static inline vsetBucket *vsetBucketFromHashtable(hashtable *ht) {
    return vsetBucketFromRawPtr(ht, VSET_BUCKET_HT);
}

static inline vsetBucket *vsetBucketFromSingle(void *ptr) {
    return ptr;
}

static inline vsetBucket *vsetBucketFromNone(void) {
    return NULL;
}

static inline vsetBucket *vsetBucketFromRax(rax *r) {
    return vsetBucketFromRawPtr(r, VSET_BUCKET_RAX);
}
```

Reply:
Sure that makes lot's of sense to me. will change it as you suggest